cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +278,495 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+                  col_id.begin());
+  auto level_boundaries = cudf::detail::make_std_vector_async(d_level_boundaries, stream);
+  // Initialize First level node's node col_id to 0
+  thrust::fill(rmm::exec_policy(stream), col_id.begin(), col_id.begin() + level_boundaries[0], 0);
+  // Initialize First level node's parent_col_id to parent_node_sentinel sentinel
+  thrust::fill(rmm::exec_policy(stream),
+               parent_col_id.begin(),
+               parent_col_id.begin() + level_boundaries[0],
+               parent_node_sentinel);
+
+#ifdef NJP_DEBUG_PRINT
+  auto print_level_data = [stream](auto stage_text,
+                                   auto level,
+                                   auto start,
+                                   auto end,
+                                   auto const& nodeII,
+                                   auto const& parent_node_idx,
+                                   auto const& parent_col_id,
+                                   auto const& node_type,
+                                   auto const& col_id) {
+    std::cout << level << stage_text << "\n";
+    auto cls  = std::array<const char*, 6>{"S", "L", "F", "R", "V", "E"};
+    auto dget = [](auto const& v, auto i) { return int(thrust::device_pointer_cast(v.data())[i]); };
+    for (auto n = start; n < end; n++)
+      printf("%3d ", dget(nodeII, n));
+    printf(" nodeII-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", dget(parent_node_idx, n));
+    printf(" parent_node_idx-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", parent_col_id.element(n, stream));
+    printf(" parent_col_id-%ld\n", level);
+    for (auto n = start; n < end; n++) {
+      auto nt = dget(node_type, n);
+      printf("%3s ",
+             nt >= NUM_NODE_CLASSES ? std::to_string(nt - NUM_NODE_CLASSES).c_str() : cls[nt]);
+    }
+    printf(" node_type-%ld\n", level);
+    for (auto n = start; n < end; n++)
+      printf("%3d ", col_id.element(n, stream));
+    printf(" col_id-%ld\n", level);
+  };
+#define PRINT_LEVEL_DATA(level, text)           \
+  print_level_data(text,                        \
+                   level,                       \
+                   level_boundaries[level - 1], \
+                   level_boundaries[level],     \
+                   scatter_indices,             \
+                   parent_indices,              \
+                   parent_col_id,               \
+                   node_type,                   \
+                   col_id);
+#else
+#define PRINT_LEVEL_DATA(level, text) ;
+#endif
+
+  // 4. Propagate parent node ids for each level.
+  // For each level,
+  //     a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+  //     b. stable sort by {parent_col_id, node_type}
+  //     c. scan sum of unique {parent_col_id, node_type}
+  //     d. scatter the col_id back to stable node_level order (using scatter_indices)
```

Comment:
This comment chunk appears 3 times in this file. Seems that would be difficult to maintain.

Reply:
removed.